### PR TITLE
Add few types and tests around generateCategoricalChart

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

So I am trying to wrap my head around how generateCategoricalChart works and in the process I added some types and tests. Not all types, just some.

Please correct me if I got something wrong - I'm new to the codebase, I have almost definitely missed something.

## Related Issue

https://github.com/recharts/recharts/discussions/3717

## Motivation and Context

Hopefully make maintenance easier

## How Has This Been Tested?

I didn't make any runtime changes - only types and tests. If the runtime code was not compatible with the types I added `@ts-expect-error` instead.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by ckifer** [comment]
These are some awesome changes that we've been meaning to do for a long time, thank you! Type safety was ignored quite a bit when this project was moved from JS to TS.

Worried a little bit about merging this as it has a lot of changes to types and this could potentially break consumers downstream who are using our types. _Most likely_ not, but you never know.

I'll finish reviewing when I can and will try to call out anything that might be risky

File: test/util/PolarUtils.spec.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import { RADIAN, degreeToRadian, getMaxRadius, polarToCartesian, radianToDegree } from '../../src/util/PolarUtils';
```
Comment (nikolasrieble) [inline test/util/PolarUtils.spec.ts:1]: To reduce the blast radius of this PR I suggest too add all tests separately. Type chances can cause pain for customers, tests never do. 

Thus I would be happy to get the tests merged in first, while paying more attention to the changes of types in the review.

File: test/util/PolarUtils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    expect(radianToDegree(rad)).toEqual(deg);
+  });
+});
+
+describe('polarToCartesian', () => {
+  it('should convert zero', () => {
+    expect(polarToCartesian(0, 0, 0, 0)).toEqual({ x: 0, y: 0 });
+  });
+
+  it('should convert 1, Pi/2', () => {
```
Comment (nikolasrieble) [inline test/util/PolarUtils.spec.ts:20]: Nit: Seems like a cases where parametrised tests could be great.

File: src/util/ChartUtils.ts [hunk 3]
```diff
@@ -847,14 +863,16 @@ export const getStackedData = (data: any, stackItems: any, offsetType: string) =
   return stack(data);
 };
 
+export type StackGroups = any;
```
Comment (nikolasrieble) [inline src/util/ChartUtils.ts:866]: I do not find that this adds value. What is the intention?

File: src/util/DataUtils.ts [hunk 4]
```diff
@@ -12,7 +12,7 @@ export const mathSign = (value: number) => {
 };
 
 export const isPercent = (value: string | number): value is `${number}%` =>
-  _.isString(value) && value.indexOf('%') === value.length - 1;
+  _.isString(value) && value.length > 1 && value.indexOf('%') === value.length - 1;
```
Comment (nikolasrieble) [inline src/util/DataUtils.ts:15]: I do believe that many of these changes will actually fix bugs in a few edge cases, and I believe them to be valuable, but I would prefer to see smaller PRs, each with more focus. 

The focus being generateCategoricalChart means it is almost the whole library.

File: test/util/ChartUtils.spec.tsx [hunk 5]
```diff
@@ -191,7 +192,7 @@ describe('getBandSizeOfAxis', () => {
 });
 
 describe('parseSpecifiedDomain', () => {
-  const domain = [20, 100];
+  const domain: [number, number] = [20, 100];
```
Comment (nikolasrieble) [inline test/util/ChartUtils.spec.tsx:195]: It would be nice to have a type 

```
type Domain = [number, number]
```

Comment (nikolasrieble) [inline test/util/ChartUtils.spec.tsx:195]: This would be a SpecifiedDomain, right? 
```suggestion
  const domain: SpecifiedDomain = [20, 100];
```

File: src/util/PolarUtils.ts [hunk 6]
```diff
@@ -39,7 +50,7 @@ export const getMaxRadius = (
  * @return {Object} Configuration
  */
 export const formatAxisMap = (
-  props: any,
+  props: FormatAxisMapProps,
```
Comment (nikolasrieble) [inline src/util/PolarUtils.ts:53]: The props are the props of axis itself, right? 

```
    Object.keys(axisObj).forEach(key => {
      axisObj[key] = formatAxisMap(props, axisObj[key], offset, key.replace('Map', ''), chartName);
    });
```suggestion
  props: FormatAxisMapProps,
```
```

File: src/util/PolarUtils.ts [hunk 7]
```diff
@@ -29,6 +29,17 @@ export const getMaxRadius = (
     Math.abs(height - (offset.top || 0) - (offset.bottom || 0)),
   ) / 2;
 
+type FormatAxisMapProps = {
```
Comment (nikolasrieble) [inline src/util/PolarUtils.ts:32]: I think instead of defining this newly, we could reuse the union of PolarAxisProps and CartesianAxisProps.

File: src/util/ChartUtils.ts [hunk 8]
```diff
@@ -574,16 +588,18 @@ export const getCoordinatesOfGrid = (ticks: Array<TickItem>, min: number, max: n
...
-  },
+  axis:
+    | null
+    | (BaseAxisProps & {
+        duplicateDomain?: any;
+        realScaleType?: 'scaleBand' | 'band' | 'point' | 'linear';
+        scale?: any;
+        axisType?: AxisType;
+        ticks?: any;
+        niceTicks?: any;
```
Comment (nikolasrieble) [inline src/util/ChartUtils.ts:599]: Wow, I have never seen this before. Will have to investigate where and how this is used.

File: src/chart/generateCategoricalChart.tsx [hunk 9]
```diff
@@ -721,6 +788,10 @@ const calculateOffset = (
   };
 };
 
+type AxisMap = {
+  [k: string]: BaseAxisProps;
+};
```
Comment (nikolasrieble) [inline src/chart/generateCategoricalChart.tsx:793]: I wonder whether this is correct. Does the axis map not hold all the props for each axis (i.e. cartesian and polar)?

**Review by nikolasrieble — COMMENTED** [review]
I am very happy about this PR, yet slightly worried about its size. Chances are that we miss some changes that might break something. 

This does add a lot of value over all, and I would be willing to take the risk, and then fix as we go. 

But I would prefer it even more if we could split this up into separate PRs

(1) Addings tests 
(2) Addings types (if possible multiple smaller ones), starting with the Util files.

**Comment by ckifer** [comment]
+1, smaller PRs will make this easier to review and easier for us to identify issues if we find them post merge

File: src/util/ChartUtils.ts (hunk 3 above)
Comment (PavelVanecek) [reply src/util/ChartUtils.ts:866]: Documentation - sort of "I don't know what this type is exactly but it's the same as over there". Bringing a bit of information into the sea of `any`.

File: test/chart/generateCategoricalChart.test.ts [hunk 10]
```diff
@@ -193,7 +202,7 @@ describe('generateCategoricalChart', () => {
             yAxisId: 0,
             tickCount: 2,
             type: 'number',
-            padding: { left: 0, right: 0 },
+            padding: { bottom: 0, top: 0 },
```
[the remark this answers is not in the document]
Comment (PavelVanecek) [reply test/chart/generateCategoricalChart.test.ts:205]: XAxisProps and YAxisProps have different paddings (which makes sense). I added a type which forced mock data change which forced the test result change.

File: test/util/ChartUtils.spec.tsx (hunk 5 above)
Comment (PavelVanecek) [reply test/util/ChartUtils.spec.tsx:195]: Quite possible - I still don't fully understand the Domain code.

**Comment by PavelVanecek** [comment]
Thx @ckifer and @nikolasrieble , I will split this up into smaller PRs.

File: src/util/ChartUtils.ts (hunk 3 above)
Comment (nikolasrieble) [reply src/util/ChartUtils.ts:866]: Fair enough. Could you please add a small comment as to what a StackGroup is?

File: src/util/PolarUtils.ts (hunk 6 above)
Comment (PavelVanecek) [reply src/util/PolarUtils.ts:53]: On closer look - these appear to be `CategoricalChartProps`.

**Comment by PavelVanecek** [comment]
I'm going to go ahead and decline this PR in favour of other smaller ones.
